### ISSUE TITLE
Invalid request times are reported by jetty metrics

### ISSUE THREAD

**Issue by mziemba** [spring-projects/spring-boot#6239]
A sample application that illustrates the issue can be found [here](https://github.com/mziemba/spring-boot-jetty-metrics)

The problem is, that when returning `CompletableFuture<String>` from a controller, I get invalid request times in jetty metrics, while for a controller returning just `String` the timings seem to be accurate.

For example, after making some requests to /test/1 (check out the sample app), I get in /metrics actuator endpoint something like:

```JSON
jettyServer.requests.snapshot.min: 1467040660251,
jettyServer.requests.snapshot.98thPercentile: 1467040660251,
jettyServer.requests.snapshot.99thPercentile: 1467040660251,
jettyServer.requests.snapshot.999thPercentile: 1467040660251,
jettyServer.requests.snapshot.max: 1467040660251,
```

while after making some requests to /test/2 (a controller returning just String), I get:
```JSON
jettyServer.requests.snapshot.min: 1,
jettyServer.requests.snapshot.98thPercentile: 37,
jettyServer.requests.snapshot.99thPercentile: 37,
jettyServer.requests.snapshot.999thPercentile: 37,
jettyServer.requests.snapshot.max: 37,
```

After quick investigation, I've found this code in com.codahale.metrics.jetty9.InstrumentedHandler:

```Java
    this.listener = new AsyncListener() {
        private long startTime;

        @Override
        public void onStartAsync(AsyncEvent event) throws IOException {
            startTime = System.currentTimeMillis();
            event.getAsyncContext().addListener(this);
        }

        ...

        @Override
        public void onComplete(AsyncEvent event) throws IOException {
            final AsyncContextState state = (AsyncContextState) event.getAsyncContext();
            final HttpServletRequest request = (HttpServletRequest) state.getRequest();
            final HttpServletResponse response = (HttpServletResponse) state.getResponse();
            updateResponses(request, response, startTime);
            if (state.getHttpChannelState().getState() != HttpChannelState.State.DISPATCHED) {
                activeSuspended.dec();
            }
        }
    };
```

After debugging it seems, that onComplete is called while onStartAsync is not. That's why startTime is set to 0.

I don't know if it is a problem with Metrics, Jetty, async servlet processing in Spring or integration of all those in Spring Boot.

I'm using Spring Boot 1.3.5.RELEASE, Metrics 3.1.2 and Jetty 9.2.16.v20160414. All the versions can be found in pom.xml in sample app.

**Comment by wilkinsona** [comment]
This appears to be a bug in `InstrumentedHandler` as it's adding the listener after async processing has already begun so it misses the `onAsyncStart` callback.

```java
        try {
            super.handle(path, request, httpRequest, httpResponse);
        } finally {
            final long now = System.currentTimeMillis();
            final long dispatched = now - start;

            activeDispatches.dec();
            dispatches.update(dispatched, TimeUnit.MILLISECONDS);

            if (state.isSuspended()) {
                if (state.isInitial()) {
                    state.addListener(listener);
                }
                activeSuspended.inc();
            } else if (state.isInitial()) {
                updateResponses(httpRequest, httpResponse, start);
            }
            // else onCompletion will handle it.
        }
```

In the above code, within the call to `super.handle(…)`, the `CompletableFuture` return value is handled by `CompletionStageReturnValueHandler` resulting in a call to `HttpServletRequest.startAsync(…)`. At this point there are no async listeners so no `onStartAsync` calls are made.

In the finally block above, a listener is then added but it's too late to receive the start event. When the request eventually completes, `IntrumentableHandler`'s listener is registered so its `onComplete` method is called.

In short this is https://github.com/dropwizard/metrics/issues/914.

@mziemba Looking at that issue, I see that you commented on it around the same time as opening this issue so I've just wasted an hour digging into this.
